test.py

```diff
@@ -683,13 +684,28 @@ class UnitTest(Test):
 
     def __init__(self, test_no: int, shortname: str, suite, args: str) -> None:
         super().__init__(test_no, shortname, suite)
-        self.path = path_to(self.mode, "test", suite.name, shortname.split('.')[0])
-        self.args = shlex.split(args) + UnitTest.standard_args
+        run_in_unshare = any(fnmatch.fnmatch(shortname, pattern)
+                             for pattern in suite.cfg.get("run_in_unshare", []))
+        cmd_args = []
+        if run_in_unshare:
+            cmd_args += ["unshare", "--map-root-user", "--net"]
+        cmd_args += [path_to(self.mode, "test", suite.name, shortname.split('.')[0])]
+        cmd_args += self._get_args()
+        cmd_args += shlex.split(args) + UnitTest.standard_args
```

Comment:
What tests require an isolated network?

There's already work (by @xtrey) to add cgroups to test.py, so if we really need this, the work should be on top of that. But I don't understand why tests need network isolation.

Reply:
there are at least two tests:

- nodetool tests: it needs a mock web server which mocks the scylla providing the HTTP API.
- s3 error injection tests: it needs a mock web server which mocks the S3 server.

and we need to run these tests in parallel, to avoid already-used-port problem. we are using `unshare`. 

@xtrey is the cgroup change already implemented?

Reply:
Why wouldn't tests run on separate networks? It's good hygiene.

Reply:
You mean different subnets? If yes, we need to implement a mechanism to allocate subnets for different tests. This allocator does not exist.

Reply:
> there are at least two tests:
> 
> * nodetool tests: it needs a mock web server which mocks the scylla providing the HTTP API.

Can we randomize its port (or listening address)?

> * s3 error injection tests: it needs a mock web server which mocks the S3 server.

Again we can randomize its port or listening address, like we do for scylla clusters.

> 
> and we need to run these tests in parallel, to avoid already-used-port problem. we are using `unshare`.
> 
> @xtrey is the cgroup change already implemented?

Reply:
> Why wouldn't tests run on separate networks? It's good hygiene.

It reduces the ability to understand what is going on.

Reply:
> > Why wouldn't tests run on separate networks? It's good hygiene.
> 
> It reduces the ability to understand what is going on.

My thinking too. 

By the way, In test/cql-pytest/run.py (see pid_to_ipd()) I implemented a scheme to pick a unique IP address based on the uniqueness of a process ID. This could have worked for all our tests without sophisticated network virtualization or even "ip reservation server", with just one exception: Tests for **changing** of IP address many times one after another - where we need many different IP addresses for the same process. A test that only needs one IP change (one old and one new) could also work with my simple scheme. Even changing many IP addresses can also work with some simple trickery - i.e., creating a do-nothing process just to ensure its process id will not be reused for a long time.

Reply:
This is the opposite of what I would do - in a separate namespace, they can all have same IPs - node 1 will be x.y.z.1, node 2 x.y.z.2, etc. - consistently across tests.

Reply:
Right now, most of the code is able to share the host network namespace. There are one or two tests that don't. It's simpler to adjust those tests than to start using unshare just for them.

How would unshare() work with a borrowed cluster? It would have to be on the cluster level, not on the test level.

Reply:
> > there are at least two tests:
> > 
> > * nodetool tests: it needs a mock web server which mocks the scylla providing the HTTP API.
> 
> Can we randomize its port (or listening address)?

The nodetool test used to use random port (and maybe even random ip) but that is not enough to avoid collisions, and we did hit a collision every now and then.

Reply:
> s3 error injection tests: it needs a mock web server which mocks the S3 server.

How does netns help to mock web server?

Reply:
> You mean different subnets? If yes, we need to implement a mechanism to allocate subnets for different tests. This allocator does not exist yet. But I doubt that it will be simpler than unshare.

That's not true, very similar allocator exists, and it's used to allocate ip:port pair for re-usable scylla clusters (`test/pylib/host_registry.py`)

Reply:
> > s3 error injection tests: it needs a mock web server which mocks the S3 server.
> 
> How does netns help to mock web server?

It helps not to collide on ips/ports when running multiple instances of the same test in parallel

Reply:
> > You mean different subnets? If yes, we need to implement a mechanism to allocate subnets for different tests. This allocator does not exist yet. But I doubt that it will be simpler than unshare.
> 
> That's not true, very similar allocator exists, and it's used to allocate ip:port pair for re-usable scylla clusters (`test/pylib/host_registry.py`)

to use it, we would have to wire a the individual tests to the allocator so that they allocate / free the ip:port resource when setting up and tearing down a fixture which uses this resource. again, i doubt that will be simpler than unshare.